src/actions/xdrViewer.js

```diff
@@ -27,3 +29,127 @@ export function fetchLatestTx(horizonBaseUrl) {
       .catch(r => dispatch({type: FETCH_SEQUENCE_FAIL, payload: r}))
   }
 }
+
+export const FETCHED_SIGNERS_SUCCESS = 'FETCHED_SIGNERS_SUCCESS';
+export const FETCHED_SIGNERS_FAIL = 'FETCHED_SIGNERS_FAIL';
+export function fetchSigners(input, type, horizonBaseUrl, networkPassphrase) {
+  return dispatch => {
+    try {
+      // Capture network for determining signature base
+      StellarSdk.Network.use(new Network(networkPassphrase))
+
+      let tx = new StellarSdk.Transaction(input);
+      const hashedSignatureBase = hash(tx.signatureBase());
+
+      // Extract all signatures on transaction
+      const signatures = extractSignatures(tx);
+
+      // Extract all source accounts from transaction (base transaction and all operations)
+      let sourceAccounts = extractSigners(tx);
+
+      // Get all signers per source account - array of promises
+      sourceAccounts = Object.keys(sourceAccounts).map(accountID => getSigner(horizonBaseUrl, accountID));
+
+      Promise.all(sourceAccounts)
+      .then(response => {
+        let allSigners = response.reduce((obj, resp) => {
+          resp.data.signers.forEach(signer => {
+            const key = signer.key;
+            if (!obj.hasOwnProperty(key)) {
+              // By nature of pre-authorized transaction, we won't ever receive a pre-auth
+              // tx hash in signatures array, so we can ignore pre-authorized transactions here.
+              switch (signer.type) {
+                case 'sha256_hash':
+                  obj[key] =   {
+                    type: signer.type,
+                    value: signer.key,
+                  };
+                  break;
+                case 'ed25519_public_key':
+                  obj[key] =   {
+                    type: signer.type,
+                    keypair: Keypair.fromPublicKey(signer.key),
+                  };
+                break;
+              }
+            }});
+          return obj;
+        }, {});
+
+        allSigners = Object.values(allSigners);
+        
+        // We are only interested in checking if each of the signatures can be verified for some valid
+        // signer for any of the source accounts in the transaction -- we are not taking into account
+        // weights, or even if this signer makes sense.
+        for (var i = 0; i < signatures.length; i ++) {
+          const sigObj = signatures[i];
+          let isValid = false;
+
+          for (var j = 0; j < allSigners.length; j ++) {
+            const signer = allSigners[j];
+            switch (signer.type) {
+              case 'sha256_hash':
+                // Convert to string since you can't compare two buffer arrays
+                isValid = signer.value.toString() === sigObj.hashx.toString();
+                break;
+              case 'ed25519_public_key':
+                isValid = signer.keypair.verify(hashedSignatureBase, sigObj.sig);
+                break;
+            }
+
+            if (isValid) {
+              break;
+            }
```

Comment:
Not sure this is invalid, at least the way the code is currently written. Here the check on `isValid` exits the loop once a valid key is found. Not only does this optimize (skipping a few rounds of checks once a valid key is seen) it also breaks since `isValid` is re-written during each pass of the internally nested loop.

Reply:
You are right, my mistake.